File: internal/content/conditions.go
// SPDX-License-Identifier: Apache-2.0

package content

import (
	"context"
	"errors"
	"fmt"
	"regexp"
	"time"
)

// SettingConditions is the setting holding the rules a field is shown under.
const SettingConditions = "conditions"

// SettingListed is the setting opting a field into the admin list as a column.
const SettingListed = "listed"

// ErrFieldHidden reports a value submitted under a field its conditions hide.
var ErrFieldHidden = errors.New("content: field is hidden")

// ErrRuleCycle reports conditions that lead back to the field they show.
var ErrRuleCycle = errors.New("content: conditions loop")

// ErrFieldReferenced reports a field a sibling's conditions read.
var ErrFieldReferenced = errors.New("content: a sibling's conditions read the field")

// decimal is the shape a number rule value takes.
var decimal = regexp.MustCompile(`^-?(0|[1-9][0-9]*)(\.[0-9]+)?$`)

// ConditionsOf returns the rules a field is shown under, none when it carries none.
func ConditionsOf(f Field) Rules {
	groups, _ := f.Settings[SettingConditions].([]any)
	rules := make(Rules, 0, len(groups))
	for _, group := range groups {
		rows, _ := group.([]any)
		held := make([]Rule, 0, len(rows))
		for _, row := range rows {
			parts, _ := row.(map[string]any)
			var rule Rule
			rule.Source, _ = parts["source"].(string)
			rule.Operator, _ = parts["operator"].(string)
			rule.Value, _ = parts["value"].(string)
			held = append(held, rule)
		}
		rules = append(rules, held)
	}
	return rules.Normalize()
}

// ConditionsSetting returns the stored form of the rules, the value to hold under SettingConditions.
func ConditionsSetting(rules Rules) []any {
	groups := []any{}
	for _, group := range rules.Normalize() {
		rows := make([]any, 0, len(group))
		for _, rule := range group {
			rows = append(rows, map[string]any{
				"source": rule.Source, "operator": rule.Operator, "value": rule.Value,
			})
		}
		groups = append(groups, rows)
	}
	return groups
}

// Listed reports whether a field is opted into the admin list.
func Listed(f Field) bool {
	listed, _ := f.Settings[SettingListed].(bool)
	return listed
}

// SourceOperators returns the operators a field of the kind offers as a rule source, none for a kind no rule reads.
func SourceOperators(kind FieldKind, multiple bool) []string {
	switch kind {
	case FieldKindText:
		return []string{OperatorIs, OperatorIsNot, OperatorContains, OperatorEmpty, OperatorNotEmpty}
	case FieldKindNumber, FieldKindDate:
		return []string{OperatorIs, OperatorIsNot, OperatorLess, OperatorGreater, OperatorEmpty, OperatorNotEmpty}
	case FieldKindBoolean:
		return []string{OperatorIs, OperatorIsNot}
	case FieldKindChoice:
		if multiple {
			return []string{OperatorContains, OperatorEmpty, OperatorNotEmpty}
		}
		return []string{OperatorIs, OperatorIsNot, OperatorEmpty, OperatorNotEmpty}
	case FieldKindMedia:
		return []string{OperatorEmpty, OperatorNotEmpty}
	default:
		return nil
	}
}

// fieldParam is the param reading one sibling field as a rule source.
type fieldParam struct {
	field Field
}

// Name returns the sibling's key.
func (p fieldParam) Name() string { return p.field.Key }

// Operators returns the operators the sibling's kind offers.
func (p fieldParam) Operators() []string {
	return SourceOperators(p.field.Kind, multipleOf(p.field))
}

// Values returns no choices, the rule builder reads the sibling itself.
func (fieldParam) Values(context.Context) ([]Choice, error) { return nil, nil }

// Stands reports why a rule value does not read as the sibling's kind, if it does not.
func (p fieldParam) Stands(operator, value string) error {
	if !needsValue(operator) {
		return nil
	}
	if value == "" {
		return valueMissing(p.field.Key)
	}
	if !p.reads(value) {
		return Refuse(ErrRuleValue, "rule_value_shape",
			fmt.Sprintf("content: rule value %q does not read as %s on %s", value, p.field.Kind, p.field.Key),
			Details{"source": p.field.Key, "value": value})
	}
	return nil
}

// reads reports whether the value is one the sibling's kind can hold.
func (p fieldParam) reads(value string) bool {
	switch p.field.Kind {
	case FieldKindNumber:
		return decimal.MatchString(value)
	case FieldKindBoolean:
		return value == "true" || value == "false"
	case FieldKindDate:
		_, err := time.Parse(dateLayout, value)
		return err == nil
	case FieldKindChoice:
		return allowsCustom(p.field) || choiceValues(p.field.Settings[SettingChoices])[value]
	default:
		return true
	}
}

// Holds reports whether the sibling's value on the screen satisfies the operator and the value.
func (p fieldParam) Holds(scr Screen, operator, value string) bool {
	return compare(operator, scr[p.field.Key], value)
}

// multipleOf reports whether a choice field takes several values.
func multipleOf(f Field) bool {
	multiple, _ := f.Settings[SettingMultiple].(bool)
	return multiple
}

// allowsCustom reports whether a choice field takes values beyond its choices.
func allowsCustom(f Field) bool {
	custom, _ := f.Settings[SettingAllowCustom].(bool)
	return custom
}

// ScopeParams returns the registry of the sibling fields a condition may read.
func ScopeParams(fields []Field) *ParamRegistry {
	params := NewParamRegistry()
	for _, f := range fields {
		if SourceOperators(f.Kind, multipleOf(f)) == nil {
			continue
		}
		p := fieldParam{field: f}
		params.held[f.Key] = p
		params.order = append(params.order, p)
	}
	return params
}

// Hidden returns the keys of the fields their conditions hide on the scope, a hidden source reading as absent.
func Hidden(fields []Field, scope Values) map[string]bool {
	params := ScopeParams(fields)
	screen := screenOf(fields, scope)
	hidden := map[string]bool{}
	ordered, looped := dependencyOrder(fields)
	for _, f := range ordered {
		rules := ConditionsOf(f)
		if len(rules) == 0 || rules.Match(screen, params) {
			continue
		}
		hidden[f.Key] = true
		delete(screen, f.Key)
	}
	for _, key := range looped {
		hidden[key] = true
	}
	return hidden
}

// screenOf returns the scope as a screen where an absent boolean reads as false.
func screenOf(fields []Field, scope Values) Screen {
	screen := make(Screen, len(scope))
	for key, value := range scope {
		screen[key] = value
	}
	for _, f := range fields {
		if f.Kind == FieldKindBoolean && screen[f.Key] == nil {
			screen[f.Key] = false
		}
	}
	return screen
}

// dependencyOrder returns the fields with every source before its dependents, and the keys a loop leaves unordered.
func dependencyOrder(fields []Field) ([]Field, []string) {
	declared := make(map[string]bool, len(fields))
	for _, f := range fields {
		declared[f.Key] = true
	}
	placed := make(map[string]bool, len(fields))
	ordered := make([]Field, 0, len(fields))
	for progress := true; progress; {
		progress = false
		for _, f := range fields {
			if placed[f.Key] || !sourcesPlaced(f, declared, placed) {
				continue
			}
			placed[f.Key] = true
			ordered = append(ordered, f)
			progress = true
		}
	}
	var looped []string
	for _, f := range fields {
		if !placed[f.Key] {
			looped = append(looped, f.Key)
		}
	}
	return ordered, looped
}

// sourcesPlaced reports whether every declared source the field's conditions read is already ordered.
func sourcesPlaced(f Field, declared, placed map[string]bool) bool {
	for _, group := range ConditionsOf(f) {
		for _, rule := range group {
			if declared[rule.Source] && !placed[rule.Source] {
				return false
			}
		}
	}
	return true
}

// Shown returns the values with every key the conditions hide taken away, at any depth.
func Shown(fields []Field, values Values) Values {
	hidden := Hidden(fields, values)
	shown := make(Values, len(values))
	for key, value := range values {
		if !hidden[key] {
			shown[key] = value
		}
	}
	for _, f := range fields {
		if f.Kind.Holds() && shown[f.Key] != nil {
			shown[f.Key] = shownInside(f, shown[f.Key])
		}
	}
	return shown
}

// shownInside returns a container's value with every key its own rules hide taken away from each row.
func shownInside(f Field, value any) any {
	if rows, listed := value.([]any); listed {
		held := make([]any, len(rows))
		for i, row := range rows {
			held[i] = shownInside(f, row)
		}
		return held
	}
	inside, held := value.(map[string]any)
	if !held {
		return value
	}
	return map[string]any(Shown(f.Fields, inside))
}

// Concealed reports the first submitted value standing under a field the scope hides, at any depth.
func Concealed(fields []Field, scope, submitted Values) error {
	hidden := Hidden(fields, scope)
	for _, f := range fields {
		value := submitted[f.Key]
		if value == nil {
			continue
		}
		if hidden[f.Key] {
			return Refuse(ErrFieldHidden, "field_hidden",
				fmt.Sprintf("%s: %s", ErrFieldHidden, f.Key), Details{"field": f.Key})
		}
		if !f.Kind.Holds() {
			continue
		}
		if err := concealedInside(f, value); err != nil {
			return err
		}
	}
	return nil
}

// concealedInside reports the first value a container hides, each row read as its own scope.
func concealedInside(f Field, value any) error {
	if rows, listed := value.([]any); listed {
		for _, row := range rows {
			if err := concealedInside(f, row); err != nil {
				return err
			}
		}
		return nil
	}
	inside, _ := value.(map[string]any)
	return Concealed(f.Fields, inside, inside)
}

// Referenced returns the sibling whose conditions read the key, and whether one does.
func Referenced(fields []Field, key string) (string, bool) {
	for _, f := range fields {
		if f.Key == key {
			continue
		}
		for _, group := range ConditionsOf(f) {
			for _, rule := range group {
				if rule.Source == key {
					return f.Key, true
				}
			}
		}
	}
	return "", false
}

// Unreferenced reports whether a sibling's conditions read the key, refusing the removal when one does.
func Unreferenced(fields []Field, key string) error {
	by, found := Referenced(fields, key)
	if !found {
		return nil
	}
	return Refuse(ErrFieldReferenced, "field_referenced",
		fmt.Sprintf("%s: %s reads %s", ErrFieldReferenced, by, key), Details{"field": key, "by": by})
}

// Stands reports whether a field's conditions close no loop and read siblings of its scope.
func Stands(siblings []Field, f Field) error {
	scope := besides(siblings, f)
	if err := Acyclic(append(scope, f)); err != nil {
		return err
	}
	return ConditionsOf(f).Validate(ScopeParams(scope))
}

// besides returns the fields other than the one named, so a field never stands as its own sibling.
func besides(fields []Field, f Field) []Field {
	kept := make([]Field, 0, len(fields))
	for _, held := range fields {
		if held.Key != f.Key {
			kept = append(kept, held)
		}
	}
	return kept
}

// Acyclic reports the first field whose conditions lead back to itself, if one does.
func Acyclic(fields []Field) error {
	_, looped := dependencyOrder(fields)
	if len(looped) == 0 {
		return nil
	}
	return Refuse(ErrRuleCycle, "rule_cycle",
		fmt.Sprintf("%s: %s", ErrRuleCycle, looped[0]), Details{"field": looped[0]})
}

File: frontend/src/content/conditions.ts
// SPDX-License-Identifier: Apache-2.0

/** The operator matching a value exactly. */
const OPERATOR_IS = '=='

/** The operator excluding a value. */
const OPERATOR_IS_NOT = '!='

/** The operator matching a value below the rule value. */
const OPERATOR_LESS = '<'

/** The operator matching a value above the rule value. */
const OPERATOR_GREATER = '>'

/** The operator matching a value holding the rule value. */
const OPERATOR_CONTAINS = 'contains'

/** The operator matching a value that is not set. */
const OPERATOR_EMPTY = 'empty'

/** The operator matching a value that is set. */
const OPERATOR_NOT_EMPTY = 'not_empty'

/** One condition a field is shown under. */
export interface ConditionRule {
	source: string
	operator: string
	value: string
}

/** The field a condition reads and the settings it carries. */
export interface ConditionField {
	key: string
	kind: string
	settings: Record<string, unknown>
}

/**
 * Returns whether the operator compares against a rule value.
 * @param operator - The operator a rule carries.
 * @returns Whether a value is needed beside it.
 */
export function needsValue(operator: string): boolean {
	return operator !== OPERATOR_EMPTY && operator !== OPERATOR_NOT_EMPTY
}

/**
 * Returns the operators a field of the kind offers as a rule source.
 * @param kind - The kind the field holds.
 * @param multiple - Whether a choice field takes several values.
 * @returns The operators offered, none for a kind no rule reads.
 */
export function operatorsFor(kind: string, multiple: boolean): string[] {
	if (kind === 'text') {
		return [OPERATOR_IS, OPERATOR_IS_NOT, OPERATOR_CONTAINS, OPERATOR_EMPTY, OPERATOR_NOT_EMPTY]
	}
	if (kind === 'number' || kind === 'date') {
		return [OPERATOR_IS, OPERATOR_IS_NOT, OPERATOR_LESS, OPERATOR_GREATER, OPERATOR_EMPTY, OPERATOR_NOT_EMPTY]
	}
	if (kind === 'boolean') {
		return [OPERATOR_IS, OPERATOR_IS_NOT]
	}
	if (kind === 'choice') {
		return multiple
			? [OPERATOR_CONTAINS, OPERATOR_EMPTY, OPERATOR_NOT_EMPTY]
			: [OPERATOR_IS, OPERATOR_IS_NOT, OPERATOR_EMPTY, OPERATOR_NOT_EMPTY]
	}
	if (kind === 'media') {
		return [OPERATOR_EMPTY, OPERATOR_NOT_EMPTY]
	}
	return []
}

/**
 * Returns whether a value stands for a field nobody filled in.
 * @param held - The value the item holds.
 * @returns Whether it reads as empty.
 */
function empty(held: unknown): boolean {
	if (held === null || held === undefined) {
		return true
	}
	if (Array.isArray(held)) {
		return held.length === 0
	}
	if (typeof held === 'object') {
		return Object.keys(held).length === 0
	}
	return held === ''
}

/**
 * Returns whether a held string, number or boolean is the rule value.
 * @param held - The value the item holds.
 * @param value - The value the rule names.
 * @returns Whether the two are the same.
 */
function equals(held: unknown, value: string): boolean {
	if (typeof held === 'string') {
		return held === value
	}
	if (typeof held === 'boolean') {
		return String(held) === value
	}
	const asked = Number(value)
	return typeof held === 'number' && value.trim() !== '' && !Number.isNaN(asked) && held === asked
}

/**
 * Returns whether a held string or number sits on the given side of the rule value.
 * @param held - The value the item holds.
 * @param value - The value the rule names.
 * @param side - Below the value as minus one, above it as one.
 * @returns Whether it sits there.
 */
function ordered(held: unknown, value: string, side: number): boolean {
	if (typeof held === 'string') {
		return side > 0 ? held > value : held < value
	}
	const asked = Number(value)
	if (typeof held !== 'number' || value.trim() === '' || Number.isNaN(asked)) {
		return false
	}
	return (held - asked) * side > 0
}

/**
 * Returns whether a held string holds the rule value inside it, or a held list holds it as a member.
 * @param held - The value the item holds.
 * @param value - The value the rule names.
 * @returns Whether it is held.
 */
function contains(held: unknown, value: string): boolean {
	if (typeof held === 'string') {
		return held.includes(value)
	}
	if (Array.isArray(held)) {
		return held.some((member) => member === value)
	}
	return false
}

/**
 * Returns whether a held value satisfies the operator and the rule value, judged by the value's shape.
 * @param operator - The operator the rule carries.
 * @param held - The value the item holds.
 * @param value - The value the rule names.
 * @returns Whether the rule holds.
 */
export function compare(operator: string, held: unknown, value: string): boolean {
	if (operator === OPERATOR_IS) {
		return equals(held, value)
	}
	if (operator === OPERATOR_IS_NOT) {
		return !equals(held, value)
	}
	if (operator === OPERATOR_LESS) {
		return ordered(held, value, -1)
	}
	if (operator === OPERATOR_GREATER) {
		return ordered(held, value, 1)
	}
	if (operator === OPERATOR_CONTAINS) {
		return contains(held, value)
	}
	if (operator === OPERATOR_EMPTY) {
		return empty(held)
	}
	return operator === OPERATOR_NOT_EMPTY && !empty(held)
}

/**
 * Returns the rules a field is shown under, none when it carries none.
 * @param field - The field to read.
 * @returns The rules, as OR groups of AND rows.
 */
export function conditionsOf(field: ConditionField): ConditionRule[][] {
	const groups = field.settings.conditions
	if (!Array.isArray(groups)) {
		return []
	}
	return groups
		.map((group) => (Array.isArray(group) ? group.map(ruleOf).filter(named) : []))
		.filter((group) => group.length > 0)
}

/**
 * Returns whether a rule names anything at all.
 * @param rule - The rule to weigh.
 * @returns Whether any of its parts was written.
 */
function named(rule: ConditionRule): boolean {
	return rule.source !== '' || rule.operator !== '' || rule.value !== ''
}

/**
 * Returns one stored rule row as the evaluator reads it.
 * @param row - The row the settings hold.
 * @returns The rule, its parts empty where the row named none.
 */
function ruleOf(row: unknown): ConditionRule {
	const held = (row ?? {}) as Record<string, unknown>
	return {
		source: typeof held.source === 'string' ? held.source : '',
		operator: typeof held.operator === 'string' ? held.operator : '',
		value: typeof held.value === 'string' ? held.value : '',
	}
}

/**
 * Returns whether a choice field takes several values.
 * @param field - The field to read.
 * @returns Whether it takes several.
 */
export function multipleOf(field: ConditionField): boolean {
	return field.settings.multiple === true
}

/**
 * Returns the scope as the evaluator reads it, where an absent boolean reads as false.
 * @param fields - The fields of the scope.
 * @param scope - The values the scope holds.
 * @returns The screen the rules read.
 */
function screenOf(fields: ConditionField[], scope: Record<string, unknown>): Record<string, unknown> {
	const screen: Record<string, unknown> = { ...scope }
	for (const field of fields) {
		if (field.kind === 'boolean' && (screen[field.key] === undefined || screen[field.key] === null)) {
			screen[field.key] = false
		}
	}
	return screen
}

/**
 * Returns whether every declared source the field's conditions read is already ordered.
 * @param field - The field to weigh.
 * @param declared - The keys the scope declares.
 * @param placed - The keys already ordered.
 * @returns Whether the field may be ordered now.
 */
function sourcesPlaced(field: ConditionField, declared: Set<string>, placed: Set<string>): boolean {
	return conditionsOf(field).every((group) =>
		group.every((rule) => !declared.has(rule.source) || placed.has(rule.source)),
	)
}

/**
 * Returns the fields with every source before its dependents, and the keys a loop leaves unordered.
 * @param fields - The fields of one scope.
 * @returns The ordered fields and the looping keys.
 */
function dependencyOrder(fields: ConditionField[]): { ordered: ConditionField[]; looped: string[] } {
	const declared = new Set(fields.map((field) => field.key))
	const placed = new Set<string>()
	const ordered: ConditionField[] = []
	for (let moved = true; moved; ) {
		moved = false
		for (const field of fields) {
			if (placed.has(field.key) || !sourcesPlaced(field, declared, placed)) {
				continue
			}
			placed.add(field.key)
			ordered.push(field)
			moved = true
		}
	}
	return { ordered, looped: fields.filter((field) => !placed.has(field.key)).map((field) => field.key) }
}

/**
 * Returns whether one rule holds on the screen, a source no sibling offers failing it.
 * @param rule - The rule to weigh.
 * @param sources - The sibling fields a rule may read.
 * @param screen - The values the rules read.
 * @returns Whether the rule holds.
 */
function ruleHolds(rule: ConditionRule, sources: Set<string>, screen: Record<string, unknown>): boolean {
	return sources.has(rule.source) && compare(rule.operator, screen[rule.source], rule.value)
}

/**
 * Returns the values with every key the conditions hide taken away, at any depth.
 * @param fields - The fields of one scope, holding the fields their containers declare.
 * @param values - The values the scope holds.
 * @returns The values to send.
 */
export function shownValues(
	fields: ConditionField[],
	values: Record<string, unknown>,
): Record<string, unknown> {
	const hidden = hiddenKeys(fields, values)
	const shown: Record<string, unknown> = {}
	for (const [key, value] of Object.entries(values)) {
		if (!hidden.has(key)) {
			shown[key] = value
		}
	}
	return shown
}

/**
 * Returns the keys of the fields their conditions hide on the scope, a hidden source reading as absent.
 * @param fields - The fields of one scope.
 * @param scope - The values the scope holds.
 * @returns The hidden keys.
 */
export function hiddenKeys(fields: ConditionField[], scope: Record<string, unknown>): Set<string> {
	const sources = new Set(
		fields.filter((field) => operatorsFor(field.kind, multipleOf(field)).length > 0).map((f) => f.key),
	)
	const screen = screenOf(fields, scope)
	const hidden = new Set<string>()
	const { ordered, looped } = dependencyOrder(fields)
	for (const field of ordered) {
		const rules = conditionsOf(field).filter((group) => group.length > 0)
		if (rules.length === 0 || rules.some((group) => group.every((rule) => ruleHolds(rule, sources, screen)))) {
			continue
		}
		hidden.add(field.key)
		delete screen[field.key]
	}
	for (const key of looped) {
		hidden.add(key)
	}
	return hidden
}
